feat(meshpassthrough): new feature documentation (#1809)
kumahq/kuma-website: app/_src/policies/meshpassthrough.md

---
title: MeshPassthrough
---

{% warning %}
This policy uses new policy matching algorithm.
{% endwarning %}

> This policy enables {{site.mesh_product_name}} to configure external traffic for specific sidecar.

Reviewer: Maybe this sounds a bit better?

```suggestion
This policy enables {{site.mesh_product_name}} to configure traffic to external destinations that is allowed to pass outside the mesh.
```

When using this policy, the [passthrough mode](/docs/{{ page.version }}/networking/non-mesh-traffic/#outgoing) flag is ignored.

## TargetRef support matrix

{% tabs targetRef useUrlFragment=false %}
{% tab targetRef Sidecar %}
| `targetRef`           | Allowed kinds         |
| --------------------- | --------------------- |
| `targetRef.kind`      | `Mesh`, `MeshSubset`  |
{% endtab %}
{% endtabs %}

To learn more about the information in this table, see the [matching docs](/docs/{{ page.version }}/policies/targetref).

## Configuration

{% warning %}

> This policy doesn't work with sidecars not using [transparent-proxy](/docs/{{ page.version }}/networking/transparent-proxying/#what-is-transparent-proxying).

Reviewer: ```suggestion
This policy doesn't work with sidecars without [transparent-proxy](/docs/{{ page.version }}/networking/transparent-proxying/#what-is-transparent-proxying).
```

{% endwarning %}

The following describes the default configuration settings of the `MeshPassthrough` policy:

> - **`enabled`**: (Optional) If true, allows traffic to all external destinations.

Reviewer: ```suggestion
- **`enabled`**: (Optional) If true, allows all traffic to pass outside.
```

> - **`appendMatch`**: List of destinations that sidecar should be able to reach. When `enabled` is `true` this list is not used. It works only when `enabled` is `false`.

Reviewer: ```suggestion
- **`appendMatch`**: List of destinations that are allowed to pass through. When `enabled` is `true` this list is not used. It only takes effect when `enabled` is `false`.
```

>   - **`type`**: Defines what type of destination should be allowed. Either `Domain`, `IP` or `CIDR`

Reviewer: ```suggestion
  - **`type`**: Defines what type of destination is allowed. Either `Domain`, `IP` or `CIDR`.
```

  - **`value`**: Destination address based on the defined `type`.

>   - **`port`**: Port at which exterbal service is available. When not defined it caches all traffic to the address.

Reviewer: ```suggestion
  - **`port`**: Port at which external destination is available. When not defined it caches all traffic to the address.
```

>   - **`protocol`**: Defines protocol of the external service.

Reviewer: ```suggestion
  - **`protocol`**: Defines protocol of the external destination.
```

    - **`tcp`**: **Can't be used when `type` is `Domain` (at TCP level we are not able to disinguish domain, in this case it is going to hijack whole traffic on this port)**.
    - **`tls`**: Should be used when TLS traffic is originated by the client application.
    - **`http`**
    - **`http2`**
    - **`grpc`**
  

> ### Wildcard domains

Reviewer: I think it's better known as [wildcard dns record](https://en.wikipedia.org/wiki/Wildcard_DNS_record).

```suggestion
### Wildcard DNS matching
```

> A `MeshPassthrough` policy allows you to create a match for a wildcard domain. You can match a subdomain and allow this traffic to go outside of the cluster.

Reviewer: I wouldn't add anything more - we already explained what the policy does

```suggestion
`MeshPassthrough` policy allows you to create a match for a wildcard subdomain.
```

{% warning %}

> Currently, partial matches are not possible. For example, the domain `*ww.example.com` will be rejected.

Reviewer: I changed it to `*w` because it's 2 characters long and it's harder to mix up with just `www`. (`*ww` vs `www`)

```suggestion
Currently, support for partial subdomain matching is not implemented. For example, a match for `*w.example.com` will be rejected.
```

{% endwarning %}

{% policy_yaml wildcard %}
```yaml
type: MeshPassthrough
name: wildcard-passthrough
mesh: default
spec:
  targetRef:
    kind: Mesh
  default:
    appendMatch:
    - type: Domain
      value: '*.cluster-1.kafka.aws.us-east-2.com'
      protocol: tls
      port: 443
```
{% endpolicy_yaml %}

### Security

> It is advised that the `MeshOperator` is responsible for the `MeshPassthrough` policy. This policy can introduce traffic outside of the mesh or even the cluster, and the `MeshOperator` should be aware of this. If you want to secure access to `MeshPassthrough` to specific services, you must choose them manually. If you rely on tags in the top-level `targetRef` you might consider securing them by one of the:
> 
> * Make sure that service owners can't freely set them (using something like [`kyverno`](https://kyverno.io/), [`OPA`](https://www.openpolicyagent.org/))
> * Accept the risk of being able to "impersonate" a passthrough label and rely on auditing.

Reviewer: Would be great if we had this https://github.com/kumahq/kuma-website/issues/927 so we can link to personas. Let's split this into multiple lines so it's easier to track in git.

```suggestion
It is advised that the `MeshOperator` is responsible for managing the `MeshPassthrough` policy.
This policy can introduce traffic outside of the mesh or even the cluster, and the `MeshOperator` should be aware of this.
If you want to restrict access to `MeshPassthrough` to specific services, you must choose them manually.
If you rely on tags in the top-level `targetRef` you might consider securing them by using one of the following techniques:

* Make sure that service owners can't freely modify them (using something like [`kyverno`](https://kyverno.io/), [`OPA`](https://www.openpolicyagent.org/) or similar)
* Accept the risk of being able to "impersonate" a passthrough label and rely on auditing to figure out any violations.
```

### Limitations

* Due to the nature of some traffic, it is not possible to combine certain protocols on the same port. You can create a `MeshPassthrough` policy that handles `tcp`, `tls`, and one of `http`, `http2`, or `grpc` traffic on the same port. Layer 7 protocols cannot be distinguished, which could introduce unexpected behavior.
* It is currently not possible to route passthrough traffic through the [zone egress](/docs/{{ page.version }}/production/cp-deployment/zoneegress/#zone-egress). However, this feature will be implemented in the future.

> * Gateways are currently not supported.

Reviewer: should we then explicitly require `proxyType: sidecar`?

Reviewer: https://github.com/kumahq/kuma/issues/10468

## Examples

### Disable passthrough for all sidecars

{% policy_yaml example1 %}
```yaml
type: MeshPassthrough
name: disable-passthrough
mesh: default
spec:
  targetRef:
    kind: Mesh
  default:
    enabled: false
```
{% endpolicy_yaml %}

> ### Enable passthrough for a subset sidecars

Reviewer: ```suggestion
### Enable passthrough for a subset of sidecars
```
